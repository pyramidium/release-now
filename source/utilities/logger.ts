// source/utilities/logger.ts
// A simple colourful console logger.

/* eslint no-console: 0 */

import chalk from 'chalk';

const info = (...message: string[]) =>
  console.info(chalk.bgMagenta.bold(' INFO '), ...message);
const warn = (...message: string[]) =>
  console.error(chalk.bgYellow.bold(' WARN '), ...message);
const error = (...message: string[]) =>
  console.error(chalk.bgRed.bold(' ERROR '), ...message);
const log = console.log;

export const logger = { info, warn, error, log };
