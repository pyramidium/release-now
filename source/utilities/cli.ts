// source/utilities/cli.ts
// CLI-related utility functions.

import { env } from 'node:process';
import chalk from 'chalk';
import chalkTemplate from 'chalk-template';
import parseArgv from 'arg';
import checkForUpdate from 'update-check';
import { resolve } from './promise.js';
import { logger } from './logger.js';
import type { Arguments } from '../types.js';

// The help text for the CLI.
const helpText = chalkTemplate`
  {bold.red release-now} - Generic CLI for creating releases on GitHub

  {bold USAGE}

    {bold $} {red release-now}
    {bold $} {red release-now} minor
    {bold $} {red release-now} --help
    {bold $} {red release-now} --version

    By default, {red release-now} will prompt you to set your desired
    version number.

    You may bump the version number by using either a {bold patch}, {bold minor} 
    or {bold major} version increment.

  {bold OPTIONS}

    -h, --help         Shows this help message

    -v, --version      Displays the current version of {red release-now}

    -d, --debug        Show debugging information
`;

/**
 * Returns the help text.
 *
 * @returns The help text shown when the `--help` option is used.
 */
export const getHelpText = (): string => helpText;

// The options the CLI accepts, and how to parse them.
const options = {
  '--help': Boolean,
  '--version': Boolean,
  '--debug': Boolean,
  // A list of aliases for the above options.
  '-h': '--help',
  '-v': '--version',
  '-d': '--debug',
};

/**
 * Parses the program's `process.argv` and returns the options and arguments.
 *
 * @returns The parsed options and arguments.
 */
export const parseArguments = (): Arguments => parseArgv(options);

/**
 * Checks for updates to this package. If an update is available, it brings it
 * to the user's notice by printing a message to the console.
 */
export const checkForUpdates = async (manifest: object): Promise<void> => {
  // Do not check for updates if the `NO_UPDATE_CHECK` variable is set.
  if (env.NO_UPDATE_CHECK) return;

  // Check for a newer version of the package.
  const [error, update] = await resolve(checkForUpdate(manifest));

  // If there is an error, throw it; and if there is no update, return.
  if (error) throw error;
  if (!update) return;

  // If a newer version is available, tell the user.
  logger.log(
    chalk.bgRed.white(' UPDATE '),
    `The latest version of \`release-now\` is ${update.latest}`,
  );
};
