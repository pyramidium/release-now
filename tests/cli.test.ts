// tests/cli.test.ts
// Tests for the CLI part of the project.

import { env } from 'node:process';
import { afterEach, describe, test, expect, vi } from 'vitest';
import manifest from '../package.json';
import { getHelpText, checkForUpdates } from '../source/utilities/cli.js';
import { logger } from '../source/utilities/logger.js';

afterEach(() => {
  vi.restoreAllMocks();
});

describe('utilities/cli', () => {
  // Make sure the help message remains the same. If we are changing the help
  // message, then make sure to run `vitest` with the `--update-snapshot` flag.
  test('render help text', () => expect(getHelpText()).toMatchSnapshot());

  // Make sure the update message is shown when the current version is not
  // the latest version.
  test('print update message when newer version exists', async () => {
    const consoleSpy = vi.spyOn(logger, 'log');

    await checkForUpdates({
      ...manifest,
      version: '0.0.0',
    });

    expect(consoleSpy).toHaveBeenCalledOnce();
    expect(consoleSpy).toHaveBeenLastCalledWith(
      expect.stringContaining('UPDATE'),
      expect.stringContaining('latest'),
    );
  });

  // Make sure the update message is not shown when the latest version is
  // running.
  test('do not print update message when on latest version', async () => {
    const consoleSpy = vi.spyOn(logger, 'log');

    await checkForUpdates({
      ...manifest,
      version: '99.99.99',
    });

    expect(consoleSpy).not.toHaveBeenCalled();
  });

  // Make sure an update check does not occur when the NO_UPDATE_CHECK env var
  // is set.
  test('do not check for updates when NO_UPDATE_CHECK is set', async () => {
    const consoleSpy = vi.spyOn(logger, 'log');

    env.NO_UPDATE_CHECK = 'true';
    await checkForUpdates({
      ...manifest,
      version: '0.0.0',
    });

    expect(consoleSpy).not.toHaveBeenCalled();
  });
});
