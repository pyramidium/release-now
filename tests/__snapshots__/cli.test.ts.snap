// Vitest Snapshot v1

exports[`utilities/cli > render help text 1`] = `
"
  [1m[31mrelease-now[39m[22m - Generic CLI for creating releases on GitHub

  [1mUSAGE[22m

    [1m$[22m [31mrelease-now[39m
    [1m$[22m [31mrelease-now[39m minor
    [1m$[22m [31mrelease-now[39m --help
    [1m$[22m [31mrelease-now[39m --version

    By default, [31mrelease-now[39m will prompt you to set your desired
    version number.

    You may bump the version number by using either a [1mpatch[22m, [1mminor[22m 
    or [1mmajor[22m version increment.

  [1mOPTIONS[22m

    -h, --help         Shows this help message

    -v, --version      Displays the current version of [31mrelease-now[39m

    -d, --debug        Show debugging information
"
`;
